### ISSUE TITLE
Extend the emptiness check to all pattern records?

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#7467]
Lifted from #7442.
The lack of eta for the following record `R` makes it fail the emptiness check.
```agda
data ⊥ : Set where

record R : Set where
  no-eta-equality; pattern
  field f : ⊥

ok : R → ⊥
ok record{ f = () }

test : R → ⊥
test ()

-- R should be empty, but that's not obvious to me
-- when checking the clause left hand side
-- test ()
```
One could argue that `test` should be accepted because it can be expanded into `ok`, just as if `R` had eta.

Note that `ok` is compiled to just `ok = erased` by the GHC backend, so there is no matching on the record, just as for `test`.

@jespercockx (as the original creator of the record emptiness check): Opinion?

**Comment by UlfNorell** [comment]
Isn't this exactly the expected semantics of the emptiness check? It doesn't do any splitting, because how much splitting would you do before giving up? Or are you expecting some special handling for single constructor/single argument data types?

**Comment by andreasabel** [comment]
Good point.  Yet the current implementation does unbounded eta-record "splitting":
https://github.com/agda/agda/blob/26ef5bcba940898ca861ea7a46304e3a30dea77b/src/full/Agda/TypeChecking/Empty.hs#L70-L75
https://github.com/agda/agda/blob/26ef5bcba940898ca861ea7a46304e3a30dea77b/src/full/Agda/TypeChecking/Empty.hs#L101-L115

Indeed, I can exploit this to make Agda loop:
```agda
data ⊥ : Set where

record R : Set where
  eta-equality; inductive
  field f : R

f : R → ⊥
f ()       -- Agda loops here
```
`R` is an empty type, but this can only be seen by recursion.  So Agda should reject this absurd pattern.

If we need a loop check here anyway, we can as well allow single-constructor data types.

**Comment by UlfNorell** [comment]
You are allowed to make recursive eta-records?? I would expect that to blow up in a lot more places than in the emptyness check.

For instance

```agda
open import Agda.Builtin.Equality

record R : Set where
  eta-equality; inductive
  field f : R

open R

loop : (let X = _) → X .f ≡ X .f → Set
loop refl = {!!}  -- type checker loops
```

My suggestion is to not expand recursive records (eta or no) in the emptiness check, rather than splitting data types and adding loop checks.

**Comment by andreasabel** [comment]
> My suggestion is to not expand recursive records (eta or no) in the emptiness check

That sounds too drastic, e.g. the following is perfectly fine:
```agda
open import Agda.Builtin.Maybe

data ⊥ : Set where

record NEList (A : Set) : Set where
  inductive
  field
    hd : A
    tl : Maybe (NEList A)

f : NEList ⊥ → ⊥
f ()
```
Agda infers `NEList` as guarded recursive record type, so it turns on eta by default.

The problem with `R` is that the user overwrote the inferred no-eta for the _unguarded_ recursive record type `R` with `eta-equality`.  Maybe this should not be allowed, at least not without warning.
We  could forbid it, and as the user to use `{-# ETA R #-}` instead, which indicates that something unsafe is going on (we do this for coinductive records).
This would need some modification to the `setEtaEquality` function:
https://github.com/agda/agda/blob/26ef5bcba940898ca861ea7a46304e3a30dea77b/src/full/Agda/TypeChecking/Positivity.hs#L171-L178
https://github.com/agda/agda/blob/26ef5bcba940898ca861ea7a46304e3a30dea77b/src/full/Agda/TypeChecking/Records.hs#L513-L516
https://github.com/agda/agda/blob/26ef5bcba940898ca861ea7a46304e3a30dea77b/src/full/Agda/TypeChecking/Monad/Base.hs#L2343-L2346
`setEtaEquality` in this case should overwrite the user directive, mark it as dead code, cast an error warning and point the user toward the ETA pragma.

> rather than splitting data types and adding loop checks

Splitting data types I have not thought about, but the loop checks might be obsolete when we limit the splitting to guarded records (then splitting stops at data types).

**Comment by UlfNorell** [comment]
I don't think I understand your argument for extending the emptiness check. As far as I see there's nothing special about non-eta records, that doesn't also apply to data types. You can replace `R` with a single-constructor data type in the OP and all the arguments are the same:

```agda
data ⊥ : Set where

data D : Set where
  mkD : ⊥ → D

ok : D → ⊥
ok (mkD ())

test : D → ⊥
test () -- fails
```
The user can rewrite `test` into `ok` and the GHC backend compiles `ok` to `ok = erased`.

**Comment by jespercockx** [comment]
I'm on @UlfNorell 's side here. We could of course think about extending the emptiness check further, but in general the problem is undecidable so I would rather have something that's at least somewhat predictable and efficient.

**Comment by andreasabel** [comment]
We can extend the check to all guarded record types without much change, only have to record whether a record type is guarded in the signature.  However, we already have the "terminating" flag for record types, so maybe this would be the right criterion (rather than guardedness).
